### ISSUE TITLE
Use ActiveModel::AttributeAssignment or similar

### ISSUE THREAD

**Issue by nicolas-brousse** [github/actionview-component#236]
### Feature request
Does `ActionView::Component::Base` could include `ActiveModel::AttributeAssignment` and use `#assign_attributes(attributes)` in `#initialize`?

Like following:

```rb
class ApplicationComponent < ActionView::Component::Base
  include ActiveModel::AttributeAssignment

  def initialize(attributes = {})
    assign_attributes(attributes) if attributes

    super()
  end
end
```

Or maybe just manage a similar assignment.

### Motivation

Let use the following component as example.

```rb
class ButtonComponent < ActionView::Component::Base
  validates :text, :href, presence: true

  attr_reader :text, :href

  def initialize(text:, href:)
    @text = text
    @href = href
  end
end
```

I notice two things:
1. We need to write twice the attribute name (in `#initialize` and add it to `.attr_reader`)
2. No matter we write `initialize(text:, href:)` or `initialize(text: nil, href: nil)` since we have validation

**Comment by joelhawksley** [comment]
@nicolas-brousse thanks for filing this!

This is definitely something that would sit nicely in an ApplicationComponent though ❤️ 🙏.

We'll be removing validations shortly in `v2.0.0`, so using `attr_reader` will no longer be as common of a pattern in components.

If you have a specific case where you think this would be useful, might you be able to open a PR with a test case?

**Comment by nicolas-brousse** [comment]
@joelhawksley thanks for your answer.

Ah good to know! No I don't think keeping `ActiveModel::Validations` in `ActionView::Component::Base` is necessary. Since we could create an `ApplicationComponent` in our projet, it sounds good enough.

So if I'm correct, we could imagine the following in the future:
```ruby
class ApplicationComponent < ActionView::Component::Base
  include ActiveModel::AttributeAssignment
  include ActiveModel::Validation

  def initialize(attributes = {})
    assign_attributes(attributes) if attributes

    super()
  end
end
```

Or even
```ruby
class ApplicationComponent < ActionView::Component::Base
  include ActiveModel::Model
end
```

**Comment by joelhawksley** [comment]
@nicolas-brousse that's correct. I'd love to see some example components using this behavior shared here: https://github.com/github/actionview-component/issues/227.

In the meantime, I'm going to close this issue ❤️

**Comment by nicolas-brousse** [comment]
@joelhawksley I'm working on demo project to do some test before we implement it on real projects (https://github.com/nicolas-brousse/avc_form_builder).

I'm doing some tests with `ApplicationComponent` and `ActiveModel::Model`. The main annoying thing I notice it is that it always need to defined an `#initialize` method in all my components, since we need it for [`#source_location`](https://github.com/github/actionview-component/blob/fc6e8a2c965757a4a8bcdc0c7dcd01af8759aa75/lib/view_component/base.rb#L148).

For example.
```ruby
class Button::ButtonComponent < ApplicationComponent
  validates :content, presence: true

  attr_accessor :class

  def initialize(*); super; end
end
```

I'll share some examples in #227 in the future. When I've gone a bit deeper with my demo project.

**Comment by joelhawksley** [comment]
@nicolas-brousse we've tried and failed to remove the initializer requirement using `const_source_location` in Ruby 2.7. 

@jonspalmer did have an idea on how to address this in https://github.com/github/actionview-component/issues/221#issuecomment-589958005, but I don't it's been worked on yet. Would you be interested in taking it on?

**Comment by nicolas-brousse** [comment]
@joelhawksley I see. I may take some time later to do some check and see if I have a proposal.

For now I work and a proof of concept for a component structure. I'll check after that.

**Comment by nicolas-brousse** [comment]
What do you think about using

```ruby
class ViewComponent::Base
  class << self
    def inherited(subclass)
      @source_location = caller.first[/^[^:]+/]
    end
  end
end
```

Or using [`ActionView::ViewPaths#append_view_path`](https://api.rubyonrails.org/classes/ActionView/ViewPaths.html#method-i-append_view_path)

**Comment by joelhawksley** [comment]
@nicolas-brousse the `caller` approach is super clever! Any idea how performant it is?

I'm wary of using `ActionView::ViewPaths#append_view_path`. I don't think it's a great idea to commingle component views with other Rails views.

**Comment by nicolas-brousse** [comment]
@joelhawksley 

> the `caller` approach is super clever! Any idea how performant it is?

For now I only did test. But I haven't any successful results. I did more tests with `caller` and have weird results. I'll try to go deeper, then if I've better results I'll do some benchmarks to tests performances.

> I'm wary of using `ActionView::ViewPaths#append_view_path`. I don't think it's a great idea to commingle component views with other Rails views.

I guess you're right with this point.  
Could we imagine having `ViewComponent::ViewPaths` with a dedicated collection? I'm not sure it's a good idea too, but it may be interesting in a context of extend distributed (by gem for example) components. And only create a custom view (a bit like `devise` do for templates).  
It's just an idea.

I opened the PR #249 to work on it.

**Comment by joelhawksley** [comment]
@nicolas-brousse if we had our own list of view paths, I could be  open to that idea. I'm excited to see what you come up with for a PR :)

**Comment by nicolas-brousse** [comment]
@joelhawksley okay, I'll make a try then. :)

**Comment by jonspalmer** [comment]
I'd done an experiment with something akin to view paths. I think it works well and IMO it would be a good way to solve this problem. Overall I'd be much happier there was a config you might have to touch vs us requiring a constructor in all ViewComponents.

As I see it there are three "common" cases to figure out:

1. Normal use of the library with components and views in app/components (98% case)
1. Including components packaged in a gem
1. 'library' components in a Rails project somewhere other than app/components


1. should work be default and is easy to arrange
2. It would be really nice if this "just worked" when you include the gem. I don't know how we can arrange it or if there is prior art for how Gems bring in  controllers/views. I guess they can use Railties to add to the view_path?
3. I don't know how common this pattern is but if you had to do some manual config to support it I'd be ok.

**Comment by nicolas-brousse** [comment]
@joelhawksley I'm agree.

I started to work on it locally yesterday and for configuration part I did the following in `ViewComponent::Engine`.

```ruby
module ViewComponent
  class Engine < Rails::Engine # :nodoc:
    config.before_configuration do |app|
      app.config.paths.add "app/components", eager_load: true
    end

    initializer "action_view_component.add_components_view_paths" do |app|
      views = app.config.paths["app/components"].existent
      unless views.empty?
        ActiveSupport.on_load(:action_view_component) { prepend_view_path(views) if respond_to?(:prepend_view_path) }
      end
    end
  end
end
```

That way it should easy to customize view paths as it is for regular view.

My tests are not working well for now, but I didn't feel comfortable with this subject since I don't really know `ActionView` system. But since it's funny to discovery it I do a try.

`ActionView::ViewPaths` store [`PathSets` scoped with the class who included `ActionView::ViewPaths`](https://github.com/rails/rails/blob/master/actionview/lib/action_view/view_paths.rb#L76-L86) so with the tests I did, I confirm that view paths are well separated.

I'm digging inside to open a PR. I'll try to make a first proof of concept without tests to validate the concept. Then I'll optimize and add tests if we are enough interested on it.

---
sources:
- https://github.com/rails/rails/blob/v6.0.2.1/railties/lib/rails/engine/configuration.rb#L51
- https://github.com/rails/rails/blob/v6.0.2.1/railties/lib/rails/engine.rb#L602-L608
